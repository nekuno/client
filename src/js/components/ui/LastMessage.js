import React, { PropTypes, Component } from 'react';
import { Link } from 'react-router';
import moment from 'moment';
import ChatUserStatusStore from '../../stores/ChatUserStatusStore';
import connectToStores from '../../utils/connectToStores';
import ReactEmoji from 'react-emoji';

function getState(props) {

    const online = ChatUserStatusStore.isOnline(props.user.id) || false;

    return {
        online
    };
}

@connectToStores([ChatUserStatusStore], getState)
export default class LastMessage extends Component {

    static propTypes = {
        user   : PropTypes.object.isRequired,
        message: PropTypes.object.isRequired,
        // Injected by @connectToStores:
        online : PropTypes.bool.isRequired
    };

    render() {
        const {message, user, online} = this.props;
        let text = ReactEmoji.emojify(message.text);
        let readed = message.readed;
        let createdAt = message.createdAt;
        let imageSrc = message.user.id === message.user_from.id ? message.user_to.photo.thumbnail.medium : message.user_from.photo.thumbnail.medium;
        let mine = message.user.id === message.user_from.id;
        let style = readed || mine ? {} : {fontWeight: 'bold', color: '#000'};

        return (
            <div className="notification">
                <Link to={`/conversations/${user.slug}`}>
                    <div className="notification-picture">
                        <img src={imageSrc}/>
                        {online ? <div className="status-online">Online</div> : ''}
                    </div>
                    <div className="notification-text">
                        <div className="notification-title">
                            <span className="notification-title-text">{user.username}</span>
                        </div>
                        <div className="notification-excerpt truncate" style={style}>
                            {text}
                        </div>
                        <div className="notification-time" title={createdAt.toLocaleString()}>
                            <span className="icon-clock"></span>&nbsp;
                            <span className="notification-time-text" style={style}>{moment(createdAt).fromNow()}</span>
                        </div>
                    </div>
                </Link>
            </div>
        );
    }
}