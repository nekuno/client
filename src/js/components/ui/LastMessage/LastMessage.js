import PropTypes from 'prop-types';
import React, { Component } from 'react';
import moment from 'moment';
import Emojify from 'react-emojione';
import RoundedImage from '../RoundedImage/RoundedImage.js';
import RoundedIcon from '../RoundedIcon/RoundedIcon.js';
import styles from './LastMessage.scss';

export default class LastMessage extends Component {

    static propTypes = {
        username          : PropTypes.string.isRequired,
        photo             : PropTypes.string.isRequired,
        slug              : PropTypes.string.isRequired,
        message           : PropTypes.object.isRequired,
        online            : PropTypes.bool.isRequired,
        proposalType      : PropTypes.string,
        onClickHandler    : PropTypes.func.isRequired,
        onUserClickHandler: PropTypes.func.isRequired
    };

    handleClick() {
        const {slug} = this.props;

        this.props.onClickHandler(slug);
    }

    handleClickUser() {
        const {slug} = this.props;

        this.props.onUserClickHandler(slug);
    }

    render() {
        const {username, photo, online, message, proposalType} = this.props;
        const createdAt = message.createdAt;
        let icon = null,
            background = null;
        
        if (proposalType) {
            switch (proposalType) {
                case 'professional-project':
                    icon = 'paperclip';
                    background = '#63CAFF';
                    break;
                case 'leisure-plan':
                    icon = 'send';
                    background = '#D380D3';
                    break;
                case 'experience-plan':
                    icon = 'compass';
                    background = '#7BD47E';
                    break;
            }
        }

        return (
            <div className={styles.lastMessage}>
                <div className={styles.photo} onClick={this.handleClickUser.bind(this)}>
                    <RoundedImage size={'small'} url={photo}/>
                    {online ? <div className={styles.statusOnline + ' small'}>Online</div> : ''}
                </div>
                <div className={styles.text} onClick={this.handleClick.bind(this)}>
                    {icon && background ?
                        <RoundedIcon icon={icon} size="small" background={background} fontSize={"12px"}/>
                        : null
                    }
                    <div className={styles.username + ' small'}>{username}</div>
                    <div className={styles.excerpt + ' small'} onClick={this.handleClick.bind(this)}>
                        <Emojify>{message.text}</Emojify>
                    </div>
                </div>

                <div className={styles.time + ' small'} title={createdAt.toLocaleString()} onClick={this.handleClick.bind(this)}>
                    <span className={styles.timeText}>{moment(createdAt).fromNow()}</span>
                </div>
            </div>
        );
    }
}