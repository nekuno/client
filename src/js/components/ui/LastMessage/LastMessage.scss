@import '../../../../scss/_partials/colors';

.last-message {
  cursor: pointer;
  min-height: 70px;
  padding: 10px 16px 5px 16px;

  .photo {
    display: inline-block;
    margin-right: 5%;
    vertical-align: top;
    width: 12%;

    .status-online {
      color: $green-experiences;
      font-size: 12px;
      line-height: 12px;
      margin-top: -7px;
      text-align: center;
      font-weight: 600;
      width: 60px;
    }
  }
  .text {
    display: inline-block;
    vertical-align: top;
    width: 80%;

    .username {
      display: inline-block;
      font-weight: 600;
      margin: auto;
      height: 35px;
      line-height: 35px;
      vertical-align: top;
    }
    .excerpt {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  }

  .time {
    color: $gray-secondary-text;
    position: absolute;
    top: 12px;
    right: 12px;
  }
}