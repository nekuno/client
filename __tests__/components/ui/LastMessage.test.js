import React from 'react';
import LastMessage from '../../../src/js/components/ui/LastMessage/LastMessage.js';
import styles from '../../../src/js/components/ui/LastMessage/LastMessage.scss';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Test LastMessage component', () => {
    const mockFn = jest.fn();
    const userMockFn = jest.fn();
    const message = {
        text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
        createdAt: new Date('01-01-1981')
    };

    const lastMessage = shallow(<LastMessage slug={'johndoe'} username={'JohnDoe'} photo={'http://via.placeholder.com/100x100'} message={message} online={true} onUserClickHandler={userMockFn} onClickHandler={mockFn}/>);

    it('should be defined', () => {
        expect(LastMessage).toBeDefined();
    });
    it('should render correctly', () => {
        expect(lastMessage).toMatchSnapshot();
    });
    it('LastMessage click event', () => {
        lastMessage.find(`.${styles.text}`).simulate('click');
        expect(mockFn).toHaveBeenCalledTimes(1);
    });
    it('LastMessage user click event', () => {
        lastMessage.find(`.${styles.photo}`).simulate('click');
        expect(userMockFn).toHaveBeenCalledTimes(1);
    });
});
