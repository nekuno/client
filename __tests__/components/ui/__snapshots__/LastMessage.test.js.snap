// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LastMessage component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LastMessage
    message={
      Object {
        "createdAt": 1980-12-31T23:00:00.000Z,
        "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      }
    }
    onClickHandler={[MockFunction]}
    onUserClickHandler={[MockFunction]}
    online={true}
    photo="http://via.placeholder.com/100x100"
    slug="johndoe"
    username="JohnDoe"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="photo"
          onClick={[Function]}
        >
          <RoundedImage
            size="small"
            url="http://via.placeholder.com/100x100"
          />
          <div
            className="statusOnline small"
          >
            Online
          </div>
        </div>,
        <div
          className="text"
          onClick={[Function]}
        >
          <div
            className="username small"
          >
            JohnDoe
          </div>
          <div
            className="excerpt small"
            onClick={[Function]}
          >
            <Emojify>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
            </Emojify>
          </div>
        </div>,
        <div
          className="time small"
          onClick={[Function]}
          title="1981-1-1 00:00:00"
        >
          <span
            className="timeText"
          >
            38 years ago
          </span>
        </div>,
      ],
      "className": "lastMessage",
      "style": Object {
        "background": null,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <RoundedImage
              size="small"
              url="http://via.placeholder.com/100x100"
            />,
            <div
              className="statusOnline small"
            >
              Online
            </div>,
          ],
          "className": "photo",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "size": "small",
              "url": "http://via.placeholder.com/100x100",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Online",
              "className": "statusOnline small",
            },
            "ref": null,
            "rendered": "Online",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            null,
            <div
              className="username small"
            >
              JohnDoe
            </div>,
            <div
              className="excerpt small"
              onClick={[Function]}
            >
              <Emojify>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
              </Emojify>
            </div>,
          ],
          "className": "text",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "JohnDoe",
              "className": "username small",
            },
            "ref": null,
            "rendered": "JohnDoe",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Emojify>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
              </Emojify>,
              "className": "excerpt small",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
              },
              "ref": null,
              "rendered": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="timeText"
          >
            38 years ago
          </span>,
          "className": "time small",
          "onClick": [Function],
          "title": "1981-1-1 00:00:00",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "38 years ago",
            "className": "timeText",
          },
          "ref": null,
          "rendered": "38 years ago",
          "type": "span",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="photo"
            onClick={[Function]}
          >
            <RoundedImage
              size="small"
              url="http://via.placeholder.com/100x100"
            />
            <div
              className="statusOnline small"
            >
              Online
            </div>
          </div>,
          <div
            className="text"
            onClick={[Function]}
          >
            <div
              className="username small"
            >
              JohnDoe
            </div>
            <div
              className="excerpt small"
              onClick={[Function]}
            >
              <Emojify>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
              </Emojify>
            </div>
          </div>,
          <div
            className="time small"
            onClick={[Function]}
            title="1981-1-1 00:00:00"
          >
            <span
              className="timeText"
            >
              38 years ago
            </span>
          </div>,
        ],
        "className": "lastMessage",
        "style": Object {
          "background": null,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <RoundedImage
                size="small"
                url="http://via.placeholder.com/100x100"
              />,
              <div
                className="statusOnline small"
              >
                Online
              </div>,
            ],
            "className": "photo",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "size": "small",
                "url": "http://via.placeholder.com/100x100",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Online",
                "className": "statusOnline small",
              },
              "ref": null,
              "rendered": "Online",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              null,
              <div
                className="username small"
              >
                JohnDoe
              </div>,
              <div
                className="excerpt small"
                onClick={[Function]}
              >
                <Emojify>
                  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
                </Emojify>
              </div>,
            ],
            "className": "text",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "JohnDoe",
                "className": "username small",
              },
              "ref": null,
              "rendered": "JohnDoe",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Emojify>
                  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
                </Emojify>,
                "className": "excerpt small",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
                },
                "ref": null,
                "rendered": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="timeText"
            >
              38 years ago
            </span>,
            "className": "time small",
            "onClick": [Function],
            "title": "1981-1-1 00:00:00",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "38 years ago",
              "className": "timeText",
            },
            "ref": null,
            "rendered": "38 years ago",
            "type": "span",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
